apps/src/lab2/views/components/AiTutor2UI.tsx

```diff
@@ -0,0 +1,114 @@
+import React, {useCallback, useEffect, useRef} from 'react';
+
+import ChatMessage from '@cdo/apps/aiComponentLibrary/chatMessage/ChatMessage';
+import {Role} from '@cdo/apps/aiComponentLibrary/chatMessage/types';
+import UserMessageEditor from '@cdo/apps/aiComponentLibrary/userMessageEditor/UserMessageEditor';
+import AiTutor2Manager from '@cdo/apps/lab2/ai/AiTutor2Manager';
+import {useAppSelector} from '@cdo/apps/util/reduxHooks';
+
+import moduleStyles from './AiTutor2UI.module.scss';
+
+interface AiTutor2UIProps {
+  allowChat?: boolean;
+  type: 'user' | 'hint';
+  question?: string;
+  getFullQuestionFromQuestion?: (
+    question: string,
+    type: 'user' | 'hint'
+  ) => string;
+}
+
+const AiTutor2UI: React.FunctionComponent<AiTutor2UIProps> = ({
+  allowChat,
+  type,
+  question,
+  getFullQuestionFromQuestion,
+}) => {
+  const currentLevelId = useAppSelector(state => state.progress.currentLevelId);
+  const scriptId = useAppSelector(state => state.lab.scriptId);
+  const channelId = useAppSelector(state => state.lab.channel?.id);
+
+  // Remember the last question asked to avoid asking it multiple times, especially
+  // as UI is re-rendered.
+  const lastQuestion = useRef<string | undefined>(undefined);
+
+  // This UI component will instantiate and use a AITutorManager.
+  const aiTutor2Manager = useRef<AiTutor2Manager | null>(null);
+  if (aiTutor2Manager.current === null) {
+    aiTutor2Manager.current = new AiTutor2Manager(
+      currentLevelId,
+      scriptId,
+      channelId
+    );
+  }
```

Comment:
Assuming we should new up a fresh instance of AI Tutor Manager every time the level changes (if it's used on multiple levels)?

Reply:
Oh yes.  What's the best mechanism?  A lifecycle notifier, or just detecting that the current level ID has changed since last render?